src/ucp/core/ucp_mm.c

```diff
@@ -106,6 +106,32 @@ static ucs_status_t ucp_memh_reg_mds(ucp_context_h context, ucp_mem_h memh,
     return UCS_OK;
 }
 
+ucs_status_t ucp_memory_type_detect_mds(ucp_context_h context, void *addr, size_t length,
+                                        ucp_mem_type_t *mem_type)
+{
+    unsigned md_index;
+
+    mem_type->md_map = 0;
+    mem_type->id = UCT_MD_MEM_TYPE_HOST;
+
+    if (!context->num_mem_type_mds) {
+        return UCS_OK;
+    }
+
+    for (md_index = 0; md_index < context->num_mds; ++md_index) {
+        if (context->tl_mds[md_index].attr.cap.mem_type != UCT_MD_MEM_TYPE_HOST) {
+            if (mem_type->id == UCT_MD_MEM_TYPE_HOST) {
+                if (uct_md_is_mem_type_owned(context->tl_mds[md_index].md, addr, length)) {
+                    mem_type->id = context->tl_mds[md_index].attr.cap.mem_type;
+                    mem_type->md_map = UCS_BIT(md_index);
+                }
+            } else if (mem_type->id == context->tl_mds[md_index].attr.cap.mem_type) {
+                mem_type->md_map |= UCS_BIT(md_index);
```

Comment:
it doesn't make sense, why need to remember all MD's which are capable of detecting memory type? just need to detect the type and that's it.. also, when creating ep, you should already pre-config which transport will be used to copy to/from every non-host memory type